### PR TITLE
feat(log-rotate): log file compression feature

### PR DESCRIPTION
**Pull request by okaybase**
### What this PR does / why we need it:

Add log file compression feature.

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

**Related issues**

fix/resolve: https://github.com/apache/apisix/issues/4699

### CODE REVIEW

File: apisix/plugins/log-rotate.lua [hunk 1]
```diff
@@ -117,6 +121,22 @@ local function rotate_file(date_str, file_type)
     local ok, msg = os.rename(file_path_org, file_path)
     core.log.info("move file from ", file_path_org, " to ", file_path,
                   " res:", ok, " msg:", msg)
+
+    if ok and enable_compression then
+        local compression_filename = new_filename .. COMPRESSION_FILE_SUFFIX
+        local cmd = string.format("cd %s && tar -zcf %s %s",
+            log_dir, compression_filename, new_filename)
+        core.log.info(cmd)
+        ok = os.execute(cmd)
```
Comment (spacewander) [inline apisix/plugins/log-rotate.lua:130]: Better to use lua-resty-shell

File: apisix/plugins/log-rotate.lua [hunk 2]
```diff
@@ -117,6 +121,22 @@ local function rotate_file(date_str, file_type)
     local ok, msg = os.rename(file_path_org, file_path)
     core.log.info("move file from ", file_path_org, " to ", file_path,
                   " res:", ok, " msg:", msg)
+
+    if ok and enable_compression then
+        local compression_filename = new_filename .. COMPRESSION_FILE_SUFFIX
+        local cmd = string.format("cd %s && tar -zcf %s %s",
```
Comment (spacewander) [inline apisix/plugins/log-rotate.lua:127]: Need to clarify in the doc that this feature requires `tar` installed.

File: t/plugin/log-rotate2.t [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            end
+
+            if not has_split_access_file or not has_split_error_file then
+               ngx.status = 500
+            else
+               ngx.status = 200
+            end
+        }
+    }
+--- error_code eval
```
Comment (spacewander) [inline t/plugin/log-rotate2.t:77]: We don't need to check if error_code == 200. It is checked by default.

File: apisix/plugins/log-rotate.lua (hunk 2 above)
Comment (okaybase) [reply apisix/plugins/log-rotate.lua:127]: okay, thanks, done

File: t/plugin/log-rotate2.t (hunk 3 above)
Comment (okaybase) [reply t/plugin/log-rotate2.t]: okay, thanks, done

File: apisix/plugins/log-rotate.lua (hunk 1 above)
Comment (okaybase) [reply apisix/plugins/log-rotate.lua:130]: yep, use lua-resty-shell need add a new package `resty.shell` to this project. currently use the common `os`.

**Review by tzssangglass — APPROVED** [review]

File: conf/config-default.yaml [hunk 4]
```diff
@@ -334,6 +334,7 @@ plugin_attr:
   log-rotate:
     interval: 3600    # rotate interval (unit: second)
     max_kept: 168     # max number of log files will be kept
+    enable_compression: false    # enable log file compression or not, default false
```
Comment (tokers) [inline conf/config-default.yaml:337]: Should tell people we use the `gzip`.

File: apisix/plugins/log-rotate.lua (hunk 2 above)
Comment (dmsolr) [reply apisix/plugins/log-rotate.lua:128]: Can we test whether "tar" is installed? If not installed, give a tip and then set `enable_compression = false`. (Just a question, if it is wrong, pls ignore this.)

File: apisix/plugins/log-rotate.lua [hunk 5]
```diff
@@ -117,6 +122,22 @@ local function rotate_file(date_str, file_type)
...
+            log_dir, compression_filename, new_filename)
+        core.log.info(cmd)
+        ok = shell.run(cmd)
```
[the remark this answers is not in the document]
Comment (okaybase) [reply apisix/plugins/log-rotate.lua]: okay, thanks, done

File: apisix/plugins/log-rotate.lua [hunk 6]
```diff
@@ -117,6 +122,22 @@ local function rotate_file(date_str, file_type)
     local ok, msg = os.rename(file_path_org, file_path)
     core.log.info("move file from ", file_path_org, " to ", file_path,
                   " res:", ok, " msg:", msg)
+
+    if ok and enable_compression then
+        local compression_filename = new_filename .. COMPRESSION_FILE_SUFFIX
+        local cmd = string.format("cd %s && tar -zcf %s %s",
+            log_dir, compression_filename, new_filename)
+        core.log.info(cmd)
```
[the remark this answers is not in the document]
Comment (okaybase) [reply apisix/plugins/log-rotate.lua:130]: okay, thanks, done

**Review by spacewander — APPROVED** [review]

**Review by tokers — APPROVED** [review]

**Pull request merged**